java/vector/src/main/java/org/apache/arrow/vector/dictionary/StructSubfieldEncoder.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.dictionary;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.complex.NonNullableStructVector;
+import org.apache.arrow.vector.complex.StructVector;
+import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.util.TransferPair;
+
+/**
+ * Sub fields encoder/decoder for Dictionary encoded {@link org.apache.arrow.vector.complex.NonNullableStructVector}.
+ */
+public class StructSubfieldEncoder {
+
+  private final List<DictionaryHashTable> hashTables;
+  private final Dictionary dictionary;
+  private final BufferAllocator allocator;
+
+  /**
+   * Construct an instance.
+   */
+  public StructSubfieldEncoder(Dictionary dictionary, BufferAllocator allocator) {
+    this.dictionary = dictionary;
+    this.allocator = allocator;
+    NonNullableStructVector dictVector = (NonNullableStructVector) dictionary.getVector();
+    hashTables = new ArrayList<>();
+    for (int i = 0; i < dictVector.getChildFieldNames().size(); i++) {
+      hashTables.add(new DictionaryHashTable(getChildVector(dictVector, i)));
+    }
+  }
+
+  private FieldVector getChildVector(NonNullableStructVector vector, int index) {
+    return vector.getChildrenFromFields().get(index);
+  }
+
+  private NonNullableStructVector cloneVector(NonNullableStructVector vector) {
```

Comment:
I think we should limit this to StructVector for now.

Reply:
Does it mean NonNullableStructVector should not call cloneVector? Is there any special reason?

Reply:
This is just my opinion, but I think we should only be dealing with FieldVectors in general.  Special casing to handle non-field vectors doesn't seem right to me.